### PR TITLE
Alerting: Apply negative matchers for route matching

### LINKED ISSUE #75196: Alerting: notification policies don't match missing label
### What happened?

A notification policy that looks for a missing label in alerts doesn't not match any alerts.

Matcher: label: `test`, operator: `=`, value: `""`

Alternative Matcher: label: `test`, operator: `!~`, value: `.+`

Alternative Matcher: label: `test`, operator: `=~`, value: `^$`

An alert without label `test` is not matched by the policy.

### What did you expect to happen?

Since the alert does not have the label `test`, the policy should match but instead only the default notification policy matches.

### Did this work before?

N/A

### How do we reproduce it?

1. Create a notification policy that matches an empty label
2. Create an alert without the tested label
3. Click `Preview routing`
4. Notice it does not match the notification policy created in step 1

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: 10.1.0

### Grafana platform?

Kubernetes

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

This fix takes negative label matchers in to account by comparing matchers with no matching labels to a label for the matcher key and an empty string value.

The changes made in this PR are direct adaptations of the [Go matching algorithms](https://github.com/prometheus/alertmanager/blob/main/pkg/labels/matcher.go) and the equivalent javascript implementation in [routing-tree.js](https://github.com/prometheus/docs/blob/main/static/routing-tree.js) from the official [routing tree editor](https://prometheus.io/webtools/alerting/routing-tree-editor/).

Supersedes https://github.com/grafana/grafana/pull/75439 

**Which issue(s) does this PR fix?**:

Fixes #75196

**Special notes for your reviewer:**

I've done my best to include some test scenarios here, but please do let me know it there's a route setup I haven't or should be testing. Plenty of edge-cases here.

### CODE REVIEW

File: public/app/features/alerting/unified/utils/notification-policies.ts [hunk 1]
```diff
@@ -11,111 +11,84 @@ import { Labels } from 'app/types/unified-alerting-dto';
...
 export interface AlertInstanceMatch {
   instance: Labels;
-  matchDetails: Map<ObjectMatcher, Labels>;
-  labelsMatch: Map<Label, LabelMatchResult>;
+  labelsMatch: LabelsMatch;
 }
 
 export interface RouteMatchResult<T extends Route> {
   route: T;
-  details: Map<ObjectMatcher, Label[]>;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/notification-policies.ts:92]: We've decided to remove the `details` here since it's not being used and adds a bit of complexity we can avoid

File: public/app/features/alerting/unified/utils/notification-policies.ts [hunk 2]
```diff
@@ -11,111 +11,84 @@ import { Labels } from 'app/types/unified-alerting-dto';
...
-
-  return matchFunction(labelValue, matcherValue);
-}
-
+// If a policy has no matchers it still can be a match, hence matchers can be empty and match can be true
+// So we cannot use null as an indicator of no match
 interface LabelMatchResult {
   match: boolean;
-  matchers: ObjectMatcher[];
+  matcher: ObjectMatcher | null;
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/utils/notification-policies.ts:18]: In the preview routing, we employ this type to indicate which matchers are associated with each label in a matching instance within a route. From a technical perspective, it's indeed possible. For instance, in the following scenario, this label matches due to multiple matchers:

In this case, the `alertname=rule5` label matches 2 different matchers:

Screenshot 2023-10-30 at 12 44 41

[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/notification-policies.ts:18]: That makes sense, from an algorithm's perspective there's only ever one matcher that decides for a label. It's either the first matcher that it doesn't match or the last matcher that does. In other words, for the following set of matchers and alert instance with `alertname=foo`

`alertname=~.*, alertname=foo` it will be the last matcher (`alertname=foo`)

`alertname!=bar, alertname=foo` the matcher will be `alertname!=foo` since it will halt evaluating other matchers after a negative match.

It looks like we're only visualizing the label that matched and not the matchers for that label so I don't think it matters all that much here anyway :)

**Review by konrad147 — APPROVED** [review]
LGTM! Nice simplification

File: public/app/features/alerting/unified/utils/notification-policies.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/utils/notification-policies.ts:18]: Thanks Gilles!
yes, you're right. I was reviewing the code and it seems we are not using it, so, this change is not going to add a regression 👍

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM!
Tested locally and it works correctly!🚀
